pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/SegmentBuildTimeLeaseExtender.java

```diff
@@ -0,0 +1,144 @@
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.google.common.util.concurrent.Uninterruptibles;
+import com.linkedin.pinot.common.protocols.SegmentCompletionProtocol;
+import com.linkedin.pinot.server.realtime.ServerSegmentCompletionProtocolHandler;
+
+
+/**
+ * Extend the lease for build time. Keep a map of segments for which lease needs to be extended.
+ * Repeat lease extension periodically as often as necessary until the segment name is removed
+ * from the map.
+ */
+public class SegmentBuildTimeLeaseExtender {
+  private static final int MAX_NUM_ATTEMPTS = 3;
+  // Always request 120s of extra build time
+  private static final int EXTRA_TIME_SECONDS = 120;
+  // Retransmit lease request 10% before lease expires.
+  private static final int REPEAT_REQUEST_PERIOD_SEC = (EXTRA_TIME_SECONDS * 9 /10);
+  private static Logger LOGGER = LoggerFactory.getLogger(SegmentBuildTimeLeaseExtender.class);
+  private static final Map<String, SegmentBuildTimeLeaseExtender> INSTANCE_TO_LEASE_EXTENDER  = new HashMap<>(1);
+
+  private ScheduledExecutorService _executor;
+  private final Map<String, Future> _segmentToFutureMap = new ConcurrentHashMap<>();
+  private final String _instanceId;
+  private final ServerSegmentCompletionProtocolHandler _protocolHandler;
+
+  public static SegmentBuildTimeLeaseExtender getLeaseExtender(final String instanceId) {
+    return INSTANCE_TO_LEASE_EXTENDER.get(instanceId);
+  }
+
+  public static synchronized  SegmentBuildTimeLeaseExtender create(final String instanceId) {
+    SegmentBuildTimeLeaseExtender leaseExtender = INSTANCE_TO_LEASE_EXTENDER.get(instanceId);
+    if (leaseExtender != null) {
+      LOGGER.warn("Instance already exists");
+    } else {
+      leaseExtender = new SegmentBuildTimeLeaseExtender(instanceId);
+      INSTANCE_TO_LEASE_EXTENDER.put(instanceId, leaseExtender);
+    }
+    return leaseExtender;
+  }
+
+  private SegmentBuildTimeLeaseExtender(String instanceId) {
+    _instanceId = instanceId;
+    _protocolHandler = new ServerSegmentCompletionProtocolHandler(_instanceId);
+    _executor = new ScheduledThreadPoolExecutor(1);
+  }
+
+  public void shutDown() {
+    if (_executor != null) {
+      _executor.shutdownNow();
+      _executor = null;
+    }
+    _segmentToFutureMap.clear();
+  }
+
+  /**
+   * Adds a segment for periodic lease request.
+   * The first lease request is sent before {@param initialBuildTimeMs} exipres. Subsequent lease requests are sent
+   * within two minutes.
+   * @param segmentId is the name of he segment that is being built
+   * @param initialBuildTimeMs is the initial time budget that SegmentCompletionManager has allocated.
+   * @param offset The offset at which this segment is being built.
+   */
+  public void addSegment(String segmentId, long initialBuildTimeMs, long offset) {
+    final long initialDelayMs = initialBuildTimeMs * 9 / 10;
```

Comment:
need to Sanity check initial build time. This cannot be negative, or larger than a specified value.
Should be sending lease requests to prevent waiting for a dead server longer than 2 minutes.